### DISCUSSION TITLE
Weird runtime crash

### DISCUSSION THREAD

**Discussion by RuiAAPeres** [pointfreeco/swift-composable-architecture#1564] [Q&A] [answered]
I have the following state, as part of a child:

```swift
  public struct State: Equatable {
    public enum Status: Equatable {
      case initial
      case loading
      case error(VitalError)
    }
    
    public var status: Status = .initial
    public var alert: AlertState<Action>?
    public var code: String = ""
    
    init(
      status: Status = .initial,
      alert: AlertState<Action>? = nil,
      code: String = ""
    ) {
      self.status = status
      self.alert = alert
      self.code = code
    }
  }
```

And the child being used in the parent:

```swift
  public var body: some ReducerProtocol<State, Action> {
    
    Reduce { state, action in
      switch action {
        case let .handleResult(credentials):
          if let credentials = credentials {
            state = .authenticated(.init(credentials: credentials))
            return Effect(value: .authenticated(.setup))
          } else {
            state = .noCode(.init())
            return .none
          }
          
        case .checkStatus:
          let value = appDependencies.credentials()
          return Effect(value: .handleResult(value))
          
        case .noCode, .authenticated:
          return .none
      }
    }
    .ifCaseLet(/State.noCode, action: /Action.noCode) {
      NoCodeReducer()
    }
    .ifCaseLet(/State.authenticated, action: /Action.authenticated) {
      AuthenticatedReducer()
    }
  }
```

If I run the above code, it crashes:

![image](https://user-images.githubusercontent.com/949286/197639087-b638e993-0e4e-46e1-80c8-af70fce0183a.png)

However, if I put the `code` as the second property of the state, like this:

```swift
    public var status: Status = .initial
    public var code: String = ""
    public var alert: AlertState<Action>?
```

I get this crash instead:

![image](https://user-images.githubusercontent.com/949286/197639228-b31c8423-343c-4bfb-bd51-b06509a0263f.png)

If I remove the `code` property completely, there's no crash.

I am using `Version 14.0.1 (14A400)` with TCA pointing to main. 

What's happening here? :)

**Comment by stephencelis** [comment, marked as answer]
@RuiAAPeres Hard to pinpoint the issue without more context, but a few things stand out:

1. What does the state look like that each case path refers to (_e.g._ what does the enum look like that contains `State.noCode` and `State.authenticated`)?
2. When you got the crash to "move," were you incrementally compiling? Was the crash consistent across cleaning your build folder and running fresh?
3. If you define your case paths from scratch using `CasePath.init(embed:extract:)`, does everything work just fine?

There are some known bugs in case path reflection. While they are generally edge cases involving existentials and enums with one case, you might be hitting one here. We'd love to track them down (or better, we'd love for Apple to support native case paths), so any more info about the types involved here would be helpful!

**Reply by RuiAAPeres** [reply]
This is so odd, but doing a clean build fixed it and I have feeling that [what I mentioned here](https://twitter.com/peres/status/1584487090787491841?s=20&t=9O-UmndgPKD_F4nzG6yTwA), was the same thing.

**Reply by stephencelis** [reply]
Now that reducers live inside types/conformances, they're more compatible with Swift's incremental build system (meaning they work with things like Injection for Xcode out of the box), but this means they are also prone to some of the issues with conditional compilation, unfortunately. Glad you got sorted!

**Comment by mluisbrown** [comment]
I get this a lot too. Don't know if it's just with TCA projects, but about once a week or so I start getting unexplained crashes like the above, and a clean build or clean build + restart Xcode solves it. Must be an Xcode / Swift compiler bug related to incremental builds.

**Comment by mbrandonw** [comment]
A few months back we found the following incremental compilation crash: https://github.com/apple/swift/issues/60706. It's reproducible in plain Swift code without TCA. The sample project shows the exact steps to produce the bug too.
